### PR TITLE
FT: objectCopy with tags

### PR DESCRIPTION
**Pull request by nicolas2bert**
objectCopy with tags

Integration PR: https://github.com/scality/Integration/pull/529

### CODE REVIEW

File: lib/api/objectCopy.js [hunk 1]
```diff
@@ -36,6 +36,12 @@ function _prepMetadata(sourceObjMD, headers, sourceIsDestination, authInfo,
     if (whichMetadata !== 'COPY' && whichMetadata !== 'REPLACE') {
         return { error: errors.InvalidArgument };
     }
+    let whichTagging = headers['x-amz-tagging-directive'];
+    // Default is COPY
+    whichTagging = whichTagging === undefined ? 'COPY' : whichTagging;
+    if (whichTagging !== 'COPY' && whichTagging !== 'REPLACE') {
+        return { error: errors.InvalidArgument };
```
Comment (LaurenSpiegel) [inline lib/api/objectCopy.js]: let's customize the error to say that the tagging directive is invalid.

File: lib/api/objectCopy.js [hunk 2]
```diff
@@ -65,6 +71,17 @@ function _prepMetadata(sourceObjMD, headers, sourceIsDestination, authInfo,
         utils.getMetaHeaders(sourceObjMD) :
         utils.getMetaHeaders(headers);
 
+    // If tagging directive is REPLACE, but you don't specify the
+    // x-amz-tagging in the request, the destination object will
+    // not have any tags.
```
Comment (LaurenSpiegel) [inline lib/api/objectCopy.js]: I think you mean -- "If tagging directive is REPLACE but the source object does not have tags, the destination object will not have any tags."

correct?

Is there a difference for us saving an empty object versus having undefined? When a get is done, will 0 be returned for tag count versus nothing?

File: tests/functional/aws-node-sdk/test/object/objectCopy.js [hunk 3]
```diff
@@ -149,6 +170,64 @@ describe('Object Copy', () => {
...
+                    checkSuccessTagging(originalTagKey, originalTagValue, done);
+                });
+        });
+
+        it('should copy an object from a source bucket to a different ' +
+            'destination bucket and copy the tag set if COPY tagging ' +
+            'directive header provided', done => {
+            s3.copyObject({ Bucket: destBucketName, Key: destObjName,
+                CopySource: `${sourceBucketName}/${sourceObjName}`,
+                TaggingDirective: 'COPY' },
```
Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectCopy.js]: can we do a test where you say 'COPY' but send new tags anyway to make sure the tags are ignored or the correct error is returned?

Comment (LaurenSpiegel) [inline tests/functional/aws-node-sdk/test/object/objectCopy.js]: let's also test when we do a copy and the origin object had no tags.

File: lib/api/objectCopy.js (hunk 2 above)
Comment (nicolas2bert) [inline lib/api/objectCopy.js]: No but I can also add: `if tagging directive is COPY but the source object does not have tags, the destination object will not have any tags.`

There is no difference between saving {} and undefined both are considered as 0 tag.
`x-amz-tagging-count` header is returned only if the count is greater than zero.

File: lib/api/objectCopy.js (hunk 1 above)
Comment (nicolas2bert) [inline lib/api/objectCopy.js]: Good idea :)

File: lib/api/objectCopy.js (hunk 2 above)
Comment (LaurenSpiegel) [inline lib/api/objectCopy.js]: ah, I think the comment would be clearer: 

// If tagging directive is REPLACE but you don't specify any
// tags in the request, the destination object will
// not have any tags.
// If tagging directive is COPY but the source object does not have tags,
// the destination object will not have any tags.

File: tests/functional/aws-node-sdk/test/object/objectCopy.js (hunk 3 above)
Comment (nicolas2bert) [inline tests/functional/aws-node-sdk/test/object/objectCopy.js]: done :)

**Comment by nicolas2bert** [comment]
@ironman-machine ok

**Comment by nicolas2bert** [comment]
E2E with its integration branch: http://ci.ironmann.io/gh/scality/Integration/13128

**Pull request merged**